From d4f23e33bf88bb4cc9d164f437f3fa36cfacb2e5 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Thu, 10 Jun 2021 03:51:06 +0300
Subject: [PATCH 251/302] media: i2c: imx296: Support forcing model through
 compatible string

The IMX296 sensor exists in two different models, a colour variant
(IMX296LQ) and a monochrome variant (IMX296LL). The device exposes the
model through registers, allowing for auto-detection with a common
"sony,imx296" compatible string. However, some camera modules disable
the ability to read the sensor model register, which makes
auto-detection impossible.

The DT bindings support model-specific "sony,imx296ll" and
"sony,imx296lq" compatible strings. When those are used, skip the
auto-detection.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
---
 drivers/media/i2c/imx296.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/imx296.c b/drivers/media/i2c/imx296.c
index 70988dd9ec13..997e17cf7a75 100644
--- a/drivers/media/i2c/imx296.c
+++ b/drivers/media/i2c/imx296.c
@@ -10,6 +10,7 @@
 #include <linux/i2c.h>
 #include <linux/module.h>
 #include <linux/mutex.h>
+#include <linux/of_device.h>
 #include <linux/regmap.h>
 #include <linux/regulator/consumer.h>
 #include <linux/slab.h>
@@ -106,6 +107,8 @@
 
 #define IMX296_SENSOR_INFO				IMX296_REG_16BIT(0x3148)
 #define IMX296_SENSOR_INFO_MONO				BIT(15)
+#define IMX296_SENSOR_INFO_IMX296LQ			0x4a00
+#define IMX296_SENSOR_INFO_IMX296LL			0xca00
 #define IMX296_S_SHSA					IMX296_REG_16BIT(0x31ca)
 #define IMX296_S_SHSB					IMX296_REG_16BIT(0x31d2)
 /*
@@ -896,6 +899,15 @@ static int imx296_identify_model(struct imx296 *imx)
 	int temp = 0;
 	int ret;
 
+	model = (uintptr_t)of_device_get_match_data(imx->dev);
+	if (model) {
+		dev_dbg(imx->dev,
+			"sensor model auto-detection disabled, forcing 0x%04x\n",
+			model);
+		imx->mono = model & IMX296_SENSOR_INFO_MONO;
+		return 0;
+	}
+
 	ret = imx296_power_on(imx);
 	if (ret < 0)
 		return ret;
@@ -1070,7 +1082,9 @@ static int imx296_remove(struct i2c_client *client)
 }
 
 static const struct of_device_id imx296_of_match[] = {
-	{ .compatible = "sony,imx296" },
+	{ .compatible = "sony,imx296", .data = NULL },
+	{ .compatible = "sony,imx296ll", .data = (void *)IMX296_SENSOR_INFO_IMX296LL },
+	{ .compatible = "sony,imx296lq", .data = (void *)IMX296_SENSOR_INFO_IMX296LQ },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, imx296_of_match);
-- 
Regards,

Laurent Pinchart

